internal/abi/fastly/httpcache_guest.go

```diff
@@ -0,0 +1,1107 @@
+//    ;;;
+//    ;;; The ABI of this function includes several unusual types of input and output parameters.
+//    ;;;
+//    ;;; The bits set in the `options_mask` input parameter describe which cache options the guest is
+//    ;;; requesting that the host provide.
+//    ;;;
+//    ;;; The `options` input parameter allows the guest to provide output parameters for
+//    ;;; pointer/length options. When the corresponding bit is set in `options_mask`, the pointer and
+//    ;;; length should be set in this record to be used by the host to provide the output.
+//    ;;;
+//    ;;; The `options_mask_out` output parameter is only used by the host to indicate the status of
+//    ;;; pointer/length data in the `options_out` record. The flag for a given pointer/length
+//    ;;; parameter is set by the host if the corresponding flag was set in `options_mask`, and the
+//    ;;; value is present in the suggested options. If the host returns a status of `$buflen`, the
+//    ;;; same set of flags will be set, but the length value of the corresponding fields in
+//    ;;; `options_out` are set to the lengths that would be required to read the full value from the
+//    ;;; host on a subsequent call.
+//    ;;;
+//    ;;; The `options_out` output parameter is where the host writes the suggested options that were
+//    ;;; requested by the guest in `options_mask`. For pointer/length data, if there was enough room
+//    ;;; to write the suggested option, the length field will contain the length of the data actually
+//    ;;; written, while the pointer field will match the input pointer.
+//    ;;;
+//    ;;; The response is not consumed.
+//    (@interface func (export "get_suggested_cache_options")
+//        (param $handle $http_cache_handle)
+//        (param $response $response_handle)
+//        (param $options_mask $http_cache_write_options_mask)
+//        (param $options (@witx pointer $http_cache_write_options))
+//        (param $options_mask_out (@witx pointer $http_cache_write_options_mask))
+//        (param $options_out (@witx pointer $http_cache_write_options))
+//        (result $err (expected (error $fastly_status)))
+//    )
+
+//go:wasmimport fastly_http_cache get_suggested_cache_options
+//go:noescape
+func fastlyHTTPCacheGetSuggestedCacheOptions(
+	h httpCacheHandle,
+	r responseHandle,
+	inMask httpCacheWriteOptionsMask,
+	inOpts prim.Pointer[httpCacheWriteOptions],
+	outMask prim.Pointer[httpCacheWriteOptionsMask],
+	outOpts prim.Pointer[httpCacheWriteOptions],
+) FastlyStatus
+
+func HTTPCacheGetSuggestedCacheOptions(h *HTTPCacheHandle, r *HTTPResponse, opts *HTTPCacheWriteOptions) (*HTTPCacheWriteOptions, error) {
+	var out HTTPCacheWriteOptions
+
+	for {
+		status := fastlyHTTPCacheGetSuggestedCacheOptions(
+			h.h,
+			r.h,
+			opts.mask,
+			prim.ToPointer(&opts.opts),
+			prim.ToPointer(&out.mask),
+			prim.ToPointer(&out.opts),
+		)
+
+		if status == FastlyStatusBufLen {
```

Comment:
Maybe I'm missing something, but it seems like this implementation will always hit this the first time around, and possibly even the second time around if both `Vary` and `SurrogateKeys` have values, since no initial call to `NewWriteBuffer` is made for either field of `out.opts`.

Reply:
Oops yeah, the Rust code has pre-allocated buffers (which I forgot) and doesn't bother to have a loop, it just has two explicit calls. I chose a loop here to match other similar Go calls but I'll allocate the buffers.

Reply:
Most of my "adaptive buffers" changes allocated a buffer of a reasonable guessed size but still had the loop in case it was too small for some writes.

Reply:
06368ef